clippy_lints/src/hashset_insert_after_contains.rs

```diff
@@ -0,0 +1,184 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::{higher, peel_hir_expr_while, SpanlessEq};
+use rustc_hir::intravisit::Visitor;
+use rustc_hir::{Block, Expr, ExprKind, Stmt, StmtKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// Checks for usage of `contains` to see if a value is not
+    /// present on `HashSet` followed by a `insert`.
+    ///
+    /// ### Why is this bad?
+    /// Using just `insert` and checking the returned `bool` is more efficient.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::collections::HashSet;
+    /// let mut set = HashSet::new();
+    /// let value = 5;
+    /// if !set.contains(&value) {
+    ///     set.insert(value);
+    ///     println!("inserted {value:?}");
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::collections::HashSet;
+    /// let mut set = HashSet::new();
+    /// let value = 5;
+    /// if set.insert(&value) {
+    ///     println!("inserted {value:?}");
+    /// }
+    /// ```
+    #[clippy::version = "1.73.0"]
+    pub HASHSET_INSERT_AFTER_CONTAINS,
+    perf,
+    "use of `contains` to see if a value is not present on a `HashSet` followed by a `insert`"
+}
+declare_lint_pass!(HashsetInsertAfterContains => [HASHSET_INSERT_AFTER_CONTAINS]);
+
+impl<'tcx> LateLintPass<'tcx> for HashsetInsertAfterContains {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+
+        let Some(higher::If {
+            cond: cond_expr,
+            then: then_expr,
+            ..
+        }) = higher::If::hir(expr)
+        else {
+            return;
+        };
+
+        let Some(contains_expr) = try_parse_contains(cx, cond_expr) else {
+            return;
+        };
+
+        if !find_insert_calls(cx, &contains_expr, then_expr) {
+            return;
+        };
+        span_lint(
+            cx,
+            HASHSET_INSERT_AFTER_CONTAINS,
+            expr.span,
+            "usage of `HashSet::insert` after `HashSet::contains`. Remove the usage of `HashSet::contains` and just call `HashSet::contains` instead",
+        );
```

Comment:
Instead of `span_lint`, I'd say we should use `span_lint_and_then`.
Here's what I'd suggest:
```suggestion
        span_lint_and_then(
            cx,
            HASHSET_INSERT_AFTER_CONTAINS,
            expr.span,
            "usage of `HashSet::insert` after `HashSet::contains`,
            |diag| {
                diag.note("`HashSet::insert` returns whether it was inserted").span_help(/* span for `contains` */, "remove the `HashSet::contains` call")
            },
        );
```